### ISSUE TITLE
GET API contains a perfect storm for DoS

### ISSUE THREAD

**Issue by Seltyk** [chaoss/augur#2463]
**Description:**
The GET endpoints of Augur's API are primed for a DoS attack. There are two problems at play:

1. Augur does not rate-limit incoming GET requests, which made it trivial to make 1000 requests:

```zsh
for i in {1..1000}; do
    curl -H 'Accept: application/json' 'https://ai.chaoss.io/api/unstable/repos' &>/dev/null &
done
```

2. Augur's main endpoints ("Get All Repo Groups", "Get All Repos", and "Get Repos in Repo Group") all return the full response from the database query. This is computationally expensive and asserts a load on the network that can take several seconds to complete. Should the database grow too large, or should too many users make these requests at once, a denial-of-service is possible.

Solving these issues is, thankfully, not too difficult. Flask extensions like [Flask-Limiter](https://flask-limiter.readthedocs.io/en/stable/) make rate limiting relatively easy to implement, and large database requests can be relaxed with pagination. As it happens, I had already intended to implement pagination in Augur's API, though until now, I had intended for it to be optional. I now believe it would be better for pagination to be required of API calls.

Here are some concerns I might expect to see in response:

- **This would be a breaking change for existing frontends** like Augur View (though, to its credit, Augur View *does* have its own speedy pagination). Personally, I think the risk of a DoS is too great to ignore just to keep existing codebases functional.
- **Stress on the backend is very low, which means this is a non-issue.** This is true, but only temporarily. Should Augur's reach be expanded, or its database grow sufficiently large, stress on the instance at ai.chaoss.io could become untennable. Right now, while the stress is low, changes to the API will affect far fewer users than if it is addressed later.

The biggest open question is: how should the pagination part of the API be implemented? At the moment, I have a proof-of-concept that reads from the URL query as `?page=N&count=M` such that any page 1≤N≤4000 contains 0≤M≤50 items. This design feels unwise to me, but I cannot discern why I feel this way. Other suggestions are welcome.

**Software versions:**
 - Augur: 0.51.0, latest `dev`
 - OS: Ubuntu 22.04.2 LTS

**Comment by sgoggins** [comment]
Thanks, @Seltyk . Rate limiting and pagination are on our road map. These are helpful points for us to keep in mind as we build them out. We of course would welcome contributions in pull request form as well! :)

**Comment by Seltyk** [comment]
Glad we're on the same page. I'll continue developing my pagination implementation for now.

**Comment by sgoggins** [comment]
@Ulincsys : Can you share your thoughts about rate limiting and pagination here. I know you did a bit of looking into this already.

**Comment by Ulincsys** [comment]
Yes, we've been discussion pagination and rate-limiting for the last several months.

Currently Augur does not keep track of requests made to the frontend or the backend. All requests are served blindly and without context as to what other requests a particular client has made within a given time period. My intent moving forward is for Augur to implement an event stream internally so that we can query the behavior patterns for clients while serving requests.

We have three levels of authentication for our endpoints at the moment:
- Unauthenticated
  - These endpoints do not require an API key or user authorization to access.
  - Most of our metrics endpoints fall into this category
- Client API key required
  - These endpoints require a `Client [key]` authorization header to access
  - The Oauth endpoints fall into this category
- User authorization required
  - These endpoints require a `Client [key], Bearer [token]` authorization header to access
  - All endpoints that handle user data fall into this category

For the authenticated endpoints, it shouldn't be too difficult to implement rate-limiting once the event system is in place, since we have identity information for the requester. We can set a limit per key per hour (like GitHub), or set limits per endpoint per key per hour. We have not decided on the particulars of the limits yet, but we're definitely open to input on what seems reasonable.

As for the unauthenticated endpoints, we need to find a system that can identify unique agents based solely on the information provided in basic HTTP requests. Is an IP address alone enough? What about multiple legitimate agents making requests from the same host?
- My thought is that we should strictly rate-limit unauthenticated endpoints based on IP address (and any other useful identifiers), and make it optional to pass an API key along with requests to these endpoints. This way, we can say that if an agent wants to make a large volume of requests to unauthenticated endpoints, then they will need to register their application so they can authenticate with an API key.

**Comment by Seltyk** [comment]
My experience is in embedded development, so rate limiting is something I haven't thought very critically of before. OpenWeatherMap, if memory serves, enforces *very* strict rate limits on unauthenticated clients, but just making an account and generating an API key significantly loosens the limits. This seems to be the model you were describing in the last paragraph, and at a glance it looks like a simple yet effective way to stop bots from hogging the server while allowing legitimate clients to leverage Augur. The threat of an IP ban isn't going to stop the most malicious attackers, as they could simply hook their bot up to a VPN, but having to make a new account may slow them down enough to discourage that behavior.

I'll pass this thread along to some web and security guys I know, their input is far more valuable than mine.

**Comment by MoralCode** [comment]
@sgoggins do you have a reference to a PR or commit that this was completed in?
